torch/lib/c10d/ProcessGroupWrapper.cpp

```diff
@@ -71,81 +93,75 @@ struct CollectiveFingerPrint {
+             << pg->getRank()
+             << " is running inconsistent collective: " << *this;
+          TORCH_CHECK(false, ss.str());
         }
       }
     }
   }
 
   // Executes and verifies the collective fingerprint.
   void verify(c10::intrusive_ptr<ProcessGroup> pg) {
-    // For collectives, all ranks should participate and call into them in the
-    // same order. Verify the same operation type is being requested.
-    int64_t op_type_int = static_cast<int64_t>(op_type_);
-    verify_num(
-        op_type_int,
-        pg,
-        c10::str(
-            "Mismatch between collective operation types across ranks.",
-            "This likely indicates an application bug where different ranks are ",
-            "calling different collectives. ",
-            "Rank ",
-            pg->getRank(),
-            " is calling collective: ",
-            opTypeToString(op_type_)));
-    // Retrieve input tensor shapes.
-    std::vector<at::Tensor> shape_tensors =
-        c10d::getTensorShapes(input_tensors_);
-    // If input_tensors_ is empty we would get no shape tensors back, but still
-    // do verification in case input_tensors_.empty() is
-    // inconsistent across ranks. In this case, sub in a single zeros tensor and
-    // ensure all ranks agree, because gloo pg does not allow collectives with
-    // empty inputs.
-    if (shape_tensors.size() == 0) {
-      shape_tensors = {at::zeros(1)};
-    }
-    // Verify dimensionality of shapes. This catches errors where tensor shapes
-    // have different dimensions such as torch.randn(2, 3) vs torch.randn(2, 3,
-    // 4). If we did not do this step and instead proceeded directly with
-    // verifying tensor shapes, we would have malformed input into allgather()
-    // and crash with an unhelpful error.
-    std::vector<at::Tensor> meta_shape_tensors =
-        c10d::getTensorShapes(shape_tensors);
-
-    verify_shapes(
-        meta_shape_tensors, /* shape_tensors_to_report= */ shape_tensors, pg);
-
-    // If all meta shapes are 0 then we can skip the below verification since
-    // it is not possible that there would be a difference. This happens only
-    // when the tensor wraps a single scalar.
-    bool skip = true;
-    for (auto & t : meta_shape_tensors) {
-      if (t.item().to<int64_t>() != 0) {
-        skip = false;
-        break;
-      }
+    at::Tensor serialized_tensor = serialize_fingerprint();
+    std::vector<at::Tensor> inp{serialized_tensor};
+    // First verify tensor shapes. This is needed because if e.g. tensor dim
+    // does not match across processes, directly verifying tensors will result
+    // in a crash during allgather, but we'd actually like to report a
+    // description about the inconsistency.
+    std::vector<at::Tensor> sp = c10d::getTensorShapes(inp);
+    verify_tensors(sp, /* tensors_to_report=*/inp, pg);
+    // Now verify consistency for the actual tensor.
+    verify_tensors(inp, inp, pg);
```

Comment:
Wouldn't this logic be simpler if we just first allgather a single int which is the size of the serialized_tensor across all processes. Then using that size we allgather all the serialized tensors. Finally, on each process we verify that all the serialized tensors are equal by just comparing them using at::eq?

Reply:
From my understanding it seems like this is basically what we're doing. `inp` is going to be a 1D tensor of size k_i on each process, and we first allgather and verify its shape. We can't directly allgather an `int` so some of the logic here exists to get it into the `tensor` formats the `allgather` API expects.

So basically the call on 122 verifies the int `k_i` is the same across all process, then we allgather the serialized tensors. And we compare them with `.equal()` which should be the same as `at::eq`. That function was originally only meant to verify shape tensors but is more generic than that now, I'll update some of the names to indicate so.